tests/imports.cpp

```diff
@@ -0,0 +1,192 @@
+      std::optional<std::filesystem::perms> perms = std::nullopt)
+  {
+    std::error_code ec = {};
+    std::filesystem::create_directories(base.path() / path, ec);
+    if (ec) {
+      return false;
+    }
+    if (perms.has_value()) {
+      return set_perms(base, path, perms.value());
+    }
+    return true;
+  }
+  [[nodiscard]] bool create_file(
+      TempDir &base,
+      const std::string &path,
+      std::string contents,
+      std::optional<std::filesystem::perms> perms = std::nullopt)
+  {
+    std::filesystem::path p(path);
+    std::filesystem::path parent = p;
+    while (parent.has_parent_path()) {
+      // These will always use the default permissions.
+      parent = parent.parent_path();
+      if (!create_dir(base, parent.string())) {
+        return false;
+      }
+    }
+
+    // Create the file at the given path.
+    auto f = base.create_file(path, false);
+    if (!f) {
+      return false;
+    }
+    if (!f->write_all(contents)) {
+      return false;
+    }
+    files.emplace_back(std::move(*f));
+    if (perms.has_value()) {
+      return set_perms(base, path, perms.value());
+    }
+    return true;
+  }
+};
+
+using checkFn = std::function<void(Imports &)>;
+
+void test(const std::string &input,
+          std::vector<std::string> &&import_paths,
+          std::variant<checkFn, std::string> check)
+{
+  BPFtrace bpftrace;
+  bpftrace.config_->unstable_import = ConfigUnstable::enable;
+  ast::ASTContext ast("stdin", input);
+  std::stringstream msg;
+  msg << "\nInput:\n" << input << "\n\nOutput:\n";
+
+  auto ok = ast::PassManager()
+                .put(ast)
+                .put(bpftrace)
+                .add(ast::AllParsePasses({}, std::move(import_paths)))
+                .run();
+  std::stringstream out;
+  ast.diagnostics().emit(out);
+  if (std::holds_alternative<checkFn>(check)) {
+    ASSERT_TRUE(ok && ast.diagnostics().ok()) << msg.str() << out.str();
+    ast::Printer printer(msg);
+    printer.visit(ast.root);
+    std::get<checkFn>(check)(ok->get<Imports>());
+  } else {
+    ASSERT_FALSE(ok && ast.diagnostics().ok()) << msg.str() << out.str();
+    EXPECT_THAT(out.str(), HasSubstr(std::get<std::string>(check)))
+        << msg.str() << out.str();
+  }
+}
+
+TEST(Imports, stdlib_works)
+{
+  std::function<void(Imports &)> fn = []([[maybe_unused]] Imports &imports) {
+    // Should be implicitly imported. As we add files to the standard library,
+    // we should check that one of each kind exists here.
+  };
+
+  // The standard library import should be implicit.
+  test(R"(BEGIN {})", {}, fn);
+
+  // Importing explicitly without any explicit file system paths should be the
+  // same result as importing the standard library implicitly.
+  test(R"(import "stdlib"; BEGIN {})", {}, fn);
+}
+
+TEST_F(ImportTest, stdlib_implicit_rules)
+{
+  auto dir = TempDir::create();
+  ASSERT_TRUE(bool(dir));
+
+  // If we have a `stdlib` directory, it should override the explicit import.
```

Comment:
i'm confused about this logic - is this how things are normally done in terms of override priority?

Reply:
Sorry, this should say `implicit`. I just tweaked the comment.

If you have an explicit `import "stdlib"` then it use the regular search for the library. If you don't have an explicit `import "stdlib"` then it will never search the filesystem. This is to prevent strange surprises and unexpected behavior.